src/lib/core/tt_ulid.c

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2010-2025, Tarantool AUTHORS, please see AUTHORS file.
+ *
+ * Redistribution and use in source and binary forms, with or
+ * without modification, are permitted provided that the following
+ * conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above
+ *    copyright notice, this list of conditions and the
+ *    following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include "tt_ulid.h"
+
+#include <random.h>
+
+#include <stdint.h>
+#include <string.h>
+#include <time.h>
+
+#include "trivia/util.h"
+
+/* Zeroed by the linker. */
+const struct tt_ulid ulid_nil;
+
+static_assert(sizeof(struct tt_ulid) == ULID_LEN, "ULID size mismatch");
+
+/* Crockford Base32 alphabet. */
+static const char ALPH[] = "0123456789ABCDEFGHJKMNPQRSTVWXYZ";
+static int8_t INV[256];
+
+/* State for monotonic generator. */
+static uint64_t last_ms;
+static uint8_t last_rand[10];
+
+static uint64_t
+realtime_ms(void)
+{
+	struct timespec ts;
+	clock_gettime(CLOCK_REALTIME, &ts);
+	return (uint64_t)ts.tv_sec * 1000ULL + ts.tv_nsec / 1000000ULL;
+}
+
+void
+tt_ulid_create(struct tt_ulid *u)
+{
+	uint64_t ms = realtime_ms();
+	if (ms < last_ms)
+		ms = last_ms;
```

Comment:
Why don't we just use monotonic clock here?

Also, we already use convenient clock methods in `core/clock.h`. Please, use them instead.

By the way, we also have `core/clock_lowres.h`. It's faster monotonic clock, but with a reduced frequency. It was designed to be used on hot path. Wouldn't it be useful here?